Adding custom bins to aggregate_downsample

aaryapatil opened the pull request:
### Description
This pull request adds custom binning to `aggregate_downsample` by allowing `time_bin_start` and `time_bin_end` to be arrays. In addition, there is now an optional  `time_bin_end` argument which can be an array as well.

Fixes #8584

pep8speaks commented:
Hello @aaryapatil :wave:! Thanks for opening this pull request, we are very grateful for your contribution! I'm a friendly :robot: that checks for style issues in this pull request, since this project follows the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guidelines. I've listed some small issues I found below, but please don't hesitate to ask if any of them are unclear!


* In the file [`astropy/timeseries/downsample.py`](https://github.com/astropy/astropy/blob/d7539301b411007cb9e05170eb078e64f535f2f2/astropy/timeseries/downsample.py):

> [Line 111:14](https://github.com/astropy/astropy/blob/d7539301b411007cb9e05170eb078e64f535f2f2/astropy/timeseries/downsample.py#L111): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 152:87](https://github.com/astropy/astropy/blob/d7539301b411007cb9e05170eb078e64f535f2f2/astropy/timeseries/downsample.py#L152): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 195:18](https://github.com/astropy/astropy/blob/d7539301b411007cb9e05170eb078e64f535f2f2/astropy/timeseries/downsample.py#L195): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`astropy/timeseries/tests/test_downsample.py`](https://github.com/astropy/astropy/blob/d7539301b411007cb9e05170eb078e64f535f2f2/astropy/timeseries/tests/test_downsample.py):

> [Line 90:62](https://github.com/astropy/astropy/blob/d7539301b411007cb9e05170eb078e64f535f2f2/astropy/timeseries/tests/test_downsample.py#L90): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

pllim commented on CHANGES.rst:
```diff
@@ -88,6 +88,8 @@ astropy.time
 astropy.timeseries
 ^^^^^^^^^^^^^^^^^^
 
+- A custom binning scheme is added in ``aggregate_downsample``. [#11266]
```
This needs to be better clarified. Maybe specify what keyword you added and what is this custom binning good for?

aaryapatil replied:
I'll add more details 🙂

pllim reviewed (commented):
Thanks! I converted your todo list to checkboxes, so you can check them off when you get to them. 😸 

I think this also needs a user facing doc on how and when to use it.

pllim commented:
Since there are todo and tests are failing, I am guessing this isn't ready for review? Please ping us again when you are ready. Thanks!

aaryapatil commented:
I have done everything other than updating user-facing docs. For some reason, some of the tests fail on the ``Python 3.7 with remote data`` and ``32-bit and parallel`` which seem to be some tests/docs errors but they work fine locally, and on other CI checks. Is this something platform/Python specific that I should be dealing with?

pllim commented:
Remote data failure -- not really platform specific. You can reproduce it locally using `--remote-data` flag with `pytest`.

image-tests-mpldev -- Yes, it is unrelated. See #11267

32-bit/parallel -- Hmm, not sure about `repr failed for <MaskedArray>: Cannot convert masked element to a Python int.`. For the parallel part, you can try to reproduce locally using `pytest-xdist` and the commands used for the job. But if that passes for you locally, you might need to use Docker to get into a 32-bit image. 

p.s. Good to have you back!

aaryapatil commented:
Thanks! I fixed the ``--remote-data`` error. I still haven't figured out the ``repr failed for <MaskedArray>: Cannot convert masked element to a Python int.` - I'll work on that this week.

P. S. Good to be back!!!

aaryapatil commented:
Seems like ``repr failed for <MaskedArray>: Cannot convert masked element to a Python int`` gets fixed when I do ``down_1['a'].data.data`` since ``down_1['a']`` is a ``MaskedColumn`` and ``down_1['a'].data`` is a ``MaskedArray`` whose data I access - this seems to be the case with the 4.2 astropy version too, so using  ``down_1['a'].data.data`` is probably a good idea.

```
In [1]: import astropy                                                                                                        

In [2]: astropy.__version__                                                                                                   
Out[2]: '4.2'

In [13]: down_1["a"]                                                                                                          
Out[13]: 
<MaskedColumn name='a' dtype='int64' length=5>
1
2
3
4
5

In [14]: down_1["a"].data                                                                                                     
Out[14]: 
masked_array(data=[1, 2, 3, 4, 5],
             mask=[False, False, False, False, False],
       fill_value=999999)

In [15]: down_1["a"].data.data                                                                                                
Out[15]: array([1, 2, 3, 4, 5])
```

```
In [1]: import astropy                                                                                                       

In [2]: astropy.__version__                                                                                                  
Out[2]: '4.3.dev443+g0369badab'

In [13]: down_1["a"]                                                                                                          
Out[13]: 
<MaskedColumn name='a' dtype='int64' length=5>
1
2
3
4
5

In [14]: down_1["a"].data                                                                                                     
Out[14]: 
masked_array(data=[1, 2, 3, 4, 5],
             mask=[False, False, False, False, False],
       fill_value=999999)

In [15]: down_1["a"].data.data                                                                                                
Out[15]: array([1, 2, 3, 4, 5])
```

However, now the array values are incorrect. I tested this with ``pytest -n=4`` with ``pytest-xdist`` and it works locally. This means maybe it is an error with the 32-bit system and how parallelization works there. I think this relates to the way ``aggregate_downsample`` is dealt with previously. Can this be reviewed so I have a better idea?

aaryapatil commented:
I've added user-facing docs. I am not entirely sure why the doc build is failing. I built the docs locally and it seems to work ``build_docs: commands succeeded
  congratulations :)``

Errors are unrelated except the ones in ``32-bit and parallel``. I tried tests with parallelization and they work which means it might be a 32 bit error. Let me know if this can be reviewed!

dhomeier commented on astropy/timeseries/downsample.py:
```diff
@@ -31,23 +32,42 @@ def reduceat(array, indices, function):
+        directly as a `~astropy.time.Time` array or as any iterable that
+        initializes the `~astropy.time.Time` class. This can also be a scalar
+        value if ``time_bin_size`` is provided. Defaults to the first
+         time in the sampled time series.
```
```suggestion
        time in the sampled time series.
```
I think this was responsible for the
`docstring of astropy.timeseries.downsample.aggregate_downsample:23: WARNING: Unexpected indentation`
in the docs build.

dhomeier commented on astropy/timeseries/downsample.py:
```diff
@@ -31,23 +32,42 @@ def reduceat(array, indices, function):
+        directly as a `~astropy.time.Time` array or as any iterable that
+        initializes the `~astropy.time.Time` class. This can also be a scalar
+        value if ``time_bin_size`` is provided. Defaults to the first
+         time in the sampled time series.
```
```suggestion
        time in the sampled time series.
```
I think this was responsible for the
`docstring of astropy.timeseries.downsample.aggregate_downsample:23: WARNING: Unexpected indentation`
in the docs build.

aaryapatil replied:
Done :) 

Quite a silly mistake - guess I should update my emacs code tests to include more PEP8 conventions.

dhomeier replied:
Do you have a flycheck plugin installed? ;-)

aaryapatil replied:
I do - I followed this more or less https://docs.astropy.org/en/latest/development/codeguide_emacs.html. But somehow it doesn't give me warnings for docstrings (or maybe I don't notice them).

dhomeier commented on astropy/timeseries/tests/test_downsample.py:
```diff
@@ -35,35 +36,96 @@ def test_timeseries_invalid():
+
+def test_time_bin_invalid():
+
+    # Make sure things crash when time_bin_* is passed incorrectly.
+
+    with pytest.raises(TypeError) as exc:
+        aggregate_downsample(TimeSeries(), time_bin_size=1)
+    assert exc.value.args[0] == ("'time_bin_size' should be a Quantity or a TimeDelta")
```
```suggestion
    # Make sure to raise the right exception when time_bin_* is passed incorrectly.

    with pytest.raises(TypeError, match=r"'time_bin_size' should be a Quantity or a TimeDelta"):
        aggregate_downsample(TimeSeries(), time_bin_size=1)
```
I think "things crash" may be more commonly understood as the interpreter segfaulting etc. ;-)

This and other exceptions could be more concisely checked using the `match` option, but I don't think we have a specific policy on that, so it remains somewhat up to taste.

dhomeier reviewed (changes requested):
Thanks a lot for this PR, which I think can be very useful for some downstream packages!
Just some minor remarks on style for now, I hope to be able to provide a fuller review next week.

dhomeier commented:
@aaryapatil since there are only the doc tests visible now, do you still have a link to the failed 32bit test?

aaryapatil commented:
@dhomeier Thanks for the review! You should be able to view the tests for the previous commit `1f63520` in Checks - I didn't change any logic in the current PR so those tests are still applicable.

dhomeier commented:
Ah, thanks, found the right pull-down menu now!
Basically binning that timeseries of exactly 1 second sampling into 1 second bins comes down to exact comparison of floating point numbers, which we generally avoid in tests. Not sure why it only popped up after your changes, but it should be possible to avoid this by setting the first value an epsilon earlier, like
`INPUT_TIME = time.Time(['2016-03-22T12:30:30.9999999999', '2016-03-22T12:30:32',`
(or the other ones a nanosecond later), so the assignment to the bins will be well-defined.

dhomeier commented on astropy/timeseries/downsample.py:
```diff
@@ -61,45 +81,83 @@ def aggregate_downsample(time_series, *, time_bin_size=None, time_bin_start=None
+    if time_bin_size is None and time_bin_end is None:
+        if time_bin_start.isscalar:
+            if n_bins is None:
+                raise TypeError("Insufficient binning arguments are provided")
```
```suggestion
                raise TypeError("With single 'time_bin_start' either 'n_bins' or 'time_bin_size|end' must be provided")
```
Already a vast improvement over the previous "time_bin_size should be a astropy.unit quantity", but even better give specific diagnostics on which input to fix.

aaryapatil commented:
@dhomeier Thanks for the review, and pointing out the 32 bit floating point error! I corrected it all and now all tests are passing 💯

aaryapatil commented:
@astrofrog tagging you here in case you want to take a look!

astrofrog commented on astropy/timeseries/downsample.py:
```diff
@@ -61,45 +81,84 @@ def aggregate_downsample(time_series, *, time_bin_size=None, time_bin_start=None
     # Use the table sorted by time
     sorted = time_series.iloc[:]
 
+    # The following part makes sure that backwards compatability is provided.
```
Maybe explain backward compatibility with what - so that this comment makes sense without the context of this pull request

astrofrog reviewed (dismissed):
This looks great! In addition to the minor comment below, one question I have is how we want to deal with overlapping bins in the case where custom bins are passed, and add a test checking the behavior in this case. We could either disallow it or make it work.

aaryapatil commented:
I could add a test for overlapping bins and check the behaviour. I am open to making it work but would like to know what kind of behaviour are we expecting? Should we bin all the values within a bin (no matter the overlapping)?

dhomeier commented:
I have no strong opinions about how to handle overlapping bins either – functionally I see no reason not to permit them, but I am a bit worried that people doing statistics with the binned data might get it wrong by effectively double-counting some data points. That's probably ultimately users' responsibility, but it should be clearly stressed in the docs, and perhaps even a warning issued in every case.

I found some other peculiar behaviour when trying to find `time_bin_start` values to guarantee a specific number of points per bin related to lightkurve/lightkurve#1042. Trying to bin a timeseries at every second point to get two points in each bin I found this:
```python
>>> ts = TimeSeries(time=Time(np.arange(10, 16), format='jd'), data=[[1., 2, 11, 12, 101, 102]], names=['flux'])
>>> timeseries.aggregate_downsample(ts, time_bin_start=ts.time[::2])
<BinnedTimeSeries length=3>
time_bin_start time_bin_size        flux      
                     s                        
    object        float64         float64     
-------------- ------------- -----------------
          10.0      172800.0 4.666666666666667
          12.0      172800.0              56.5
          14.0       86400.0             102.0
>>> timeseries.aggregate_downsample(ts, time_bin_start=ts.time[::2] - 0.1 * u.ns)
<BinnedTimeSeries length=3>
  time_bin_start    time_bin_size     flux 
                          s                
      object           float64      float64
------------------ ---------------- -------
 9.999999999999998         172800.0     1.5
11.999999999999998         172800.0    11.5
13.999999999999998 86400.0000000001   101.5
```
So in the first case the bins contain data points [[0, 1, 2], [3, 4], [5,]].
This could of course be written of as normal roundoff errors on floating point values, but it seems the points _within_ numerical precision of a bin edge _always_ end up in the previous bin. Could not find where this is evaluated, but it would be nice to automatically allow some tolerance here, though maybe it's a bug (or feature?) within `BinnedTimeSeries` that is not easily changed. The workaround is easy enough...

dhomeier commented:
@aaryapatil for my part something like adding a check and warning for `np.any(time_bin_end[:-1] < time_bin_start[1:])` should do for the overlapping bins; @astrofrog do you think an additional test is needed?
Would be great to get this in before the 4.3 feature freeze tomorrow/today.

aaryapatil commented:
@dhomeier Sorry for the delay! I have added a check for overlapping bins with a warning. I have also added a test for this; overlapping bins are allowed and all the values within a bin are used (no matter the overlapping). Now the only thing left is to add this in the docs to make sure users know this as well as to fix the issue `points within numerical precision of a bin edge always end up in the previous bin`. We could potentially create issues for these, and merge this PR. This was we can make sure this goes in with the feature freeze. What do you think?

dhomeier commented on astropy/timeseries/tests/test_downsample.py:
```diff
@@ -35,35 +37,104 @@ def test_timeseries_invalid():
+                                                "can lead to double-counting of data during binning."):
+        down_overlap_bins = aggregate_downsample(ts, time_bin_start=Time(['2016-03-22T12:30:31.000',
+                                                                          '2016-03-22T12:30:33.000']),
+                                                 time_bin_end=Time(['2016-03-22T12:30:34.000',
```
```suggestion
                                                 time_bin_end=Time(['2016-03-22T12:30:33.999,
```
needs to stay `< '2016-03-22T12:30:34.000'` since `ts.time[3]` changed above.

dhomeier reviewed (changes requested):
> add this in the docs to make sure users know this as well as to fix the issue `points within numerical precision of a bin edge always end up in the previous bin`. We could potentially create separate issues for these which can be dealt with later, and merge this PR as is; that way we can make sure this goes in with the feature freeze. What do you think?

Thanks for the update; I had not looked into this in a while as well!
I think I have tracked the bin edge issue down (not really on of precision after all) – see the suggestion to update `indices`.
So that is rather crucial; tests could be updated/reverted again accordingly. Unfortunately this introduces in turn some conflicts with overlapping bins, see comment to the suggestion.
The remainder is mainly a more radical suggestion to change the whole bin assignment to directly use the time values and skip the whole conversion to relative times and seconds. I don't know if this might cause any precision problems, but I'd imagine that the whole differences calculation and conversion to seconds hardly improves accuracy.
I've tested this with our oldest supported numpy as well; since the previous implementation already had the conversion to seconds, the reason might have been that older numpy versions could not handle `ts_sorted.time` directly, but with 1.17 there seems to be no problem. But if you are uneasy about these changes, feel free to ignore them for this PR.

aaryapatil commented on astropy/timeseries/downsample.py:
```diff
@@ -61,45 +81,93 @@ def aggregate_downsample(time_series, *, time_bin_size=None, time_bin_start=None
+        time_bin_end = Time(time_bin_end)
 
     # Use the table sorted by time
     sorted = time_series.iloc[:]
```
Good catch! I was concerned about this too when I was making the changes, but decided to keep it as is since it was part of the original code. Will change :)

aaryapatil commented on astropy/timeseries/downsample.py:
```diff
@@ -61,45 +81,93 @@ def aggregate_downsample(time_series, *, time_bin_size=None, time_bin_start=None
+
+    # Check for overlapping bins, and warn if they are present
+    if time_bin_end is not None:
+        if not time_bin_end.isscalar and not time_bin_start.isscalar:
+            if np.any(time_bin_start[1:] < time_bin_end[:-1]):
+                warnings.warn("Overlapping bins should be avoided since they "
+                              "can lead to double-counting of data during binning.",
+                              AstropyUserWarning)
```
True, I separated the checks for readability but combining them seems good.

aaryapatil commented on astropy/timeseries/downsample.py:
```diff
@@ -61,45 +81,93 @@ def aggregate_downsample(time_series, *, time_bin_size=None, time_bin_start=None
+    # Assign `n_bins` since it is needed later
+    if n_bins is None:
+        n_bins = len(bin_start)
+
+    # Find the relative time since the start time, in seconds
+    relative_time_sec = (sorted.time - bin_start[0]).sec
+    # Duration of binned timeseries in seconds
+    bin_duration = (bin_end[-1] - bin_start[0]).sec
+
+    bin_start_sec = (bin_start - bin_start[0]).sec
+    bin_end_sec = (bin_end - bin_start[0]).sec
```
True. I was a bit concerned this implementation wasn't the most efficient both in terms of lines of code and extra memory/computation potentially adding precision issues, so this change makes sense. I was following the code style of this file but there's always room for improvement! 👩‍💻

aaryapatil commented on astropy/timeseries/downsample.py:
```diff
@@ -61,45 +81,93 @@ def aggregate_downsample(time_series, *, time_bin_size=None, time_bin_start=None
 
-    # Create new binned time series
-    binned = BinnedTimeSeries(time_bin_start=bins[:-1], time_bin_end=bins[-1])
+    # Figure out which bin each row falls in by sorting with respect
+    # to the bin end times
+    indices = np.searchsorted(bin_end_sec, relative_time_sec[keep])
```
Nice! Thanks for the suggestion. I was looking into this part earlier since I realized that's where the issue was arising. However, couldn't think of something on the fly that would work for all different cases without having to handle each one separately. 

Maybe `np.ndenumerate` might work for for assigning _indices_. I will try that.

I'll also try to think of an elegant solution to handle overlapping bins separately.

dhomeier commented:
When you resolve the conflicts or rebase, make sure to add the `['time']` xref from #11595 to the `time_bin_size` description; the rest should be straightforward.

eteq commented:
Since this hasn't made the feature freeze cutoff and is not a bug/docfix I'm re-milestoning to 5.0

(but note it looks *very* close and I'm not sure @aaryapatil knew the feature freeze deadline was imminent, so if you can get it in ASAP we might squeeze it into 4.3 with an explicit backport.)

aaryapatil commented:
I can try to get this done within the next few hours! I thought I had already passed the deadline so I didn’t rush.

aaryapatil commented:
I did a very stupid error and rebased with master. I will try to fix it up real quick.

dhomeier commented:
Thanks a lot, this looks great! I hope this can still be backported, would be really good to have it in 4.3.
Unfortunately I don't really know what to make of the
`datetime.datetime(2021, 12, 28, 0, 0) = <class 'erfa.helpers.leap_seconds'>._expires`
failures and if there even is anything that can be done about it on this side.
Have you rebased for the last commits?

aaryapatil commented:
I rebased everything and since then these weird errors show up. The tests run and pass locally though.

pllim reviewed (changes requested):
There is `[ci skip]` or `[skip ci]` in the commit message, so the full CI isn't running. The full CI needs to run and pass before we can accept this. Thanks!

aaryapatil commented:
That makes sense! I will reword the commit messages. The squash must have kept [skip ci].

dhomeier commented:
It looks like somehow the changes from #11638 had not been included fully or consistently; though I don't find anything here even touching those files. Let's give it another try without [ci skip].

dhomeier commented on astropy/timeseries/tests/test_downsample.py:
```diff
@@ -6,12 +6,17 @@
-                   '2016-03-22T12:30:33', '2016-03-22T12:30:34'])
+                   '2016-03-22T12:30:33', '2016-03-22T12:30:34',
+                   '2016-03-22T12:30:35'])
+ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5]], names=['a'])
+ts_units = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5] * u.count], names=['a'])
```
```suggestion
```
Ah, these apparently need to be moved inside the function; see files changes in #11638!

dhomeier commented on astropy/timeseries/tests/test_downsample.py:
```diff
@@ -33,38 +38,115 @@ def test_timeseries_invalid():
-def test_downsample():
-    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4]], names=['a'])
-    ts_units = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4] * u.count], names=['a'])
+    # Make sure to raise the right exception when time_bin_* is passed incorrectly.
```
```suggestion
    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4]], names=['a'])
    ts_units = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4] * u.count], names=['a'])
    # Make sure to raise the right exception when time_bin_* is passed incorrectly.
```
... and put them back in here.

dhomeier reviewed (changes requested):
I think that was one crucial change from the leap second PR that got merged in the wrong way.

aaryapatil commented:
Thanks! That worked. The 32 bit error is upon us again, but I think that should be fixed by maybe reducing the start times by a minute difference.

dhomeier commented on astropy/timeseries/tests/test_downsample.py:
```diff
@@ -6,12 +6,15 @@
 INPUT_TIME = Time(['2016-03-22T12:30:31', '2016-03-22T12:30:32',
-                   '2016-03-22T12:30:33', '2016-03-22T12:30:34'])
+                   '2016-03-22T12:30:33', '2016-03-22T12:30:34',
+                   '2016-03-22T12:30:35'])
```
```suggestion
                   '2016-03-22T12:30:35']).ns
# Avoid precision problems with floating-point comparisons on 32bit
if np.finfo(float).dtype == np.float32:
    INPUT_TIME += 1 * u.ns
```
maybe try this instead...

aaryapatil replied:
I tried both `INPUT_TIME += 1 * u.ns` and `INPUT_TIME[1:] += 1 * u.ns` and they don't work.
I'll see if subtracting works. Maybe the code for getting precedence of start times is causing extra precision errors?

dhomeier replied:
Good catch with `INPUT_TIME[1:]` – I had overlooked that this is testing the `time_bin_size` case, so incrementing `INPUT_TIME[0]` would shift all start values as well...
Maybe instead of the above just define something like
```python
    if np.finfo(float).dtype == np.float32:
        time_bin_incr = (1 - 1e-6) * u.s
    else:
        time_bin_incr = 1 * u.s
```
and then proceed with 1 *, 2 * `time_bin_incr` etc.
Lastly, I'd be entirely fine with marking anything that still would not pass `skipif` or `xfail` for 32bit.
Just noticed a comment about `Time` in second representation not fitting into 32bit integers; no idea if this might be in any way relevant here, just there might be issues beyond simple floating point precision...

dhomeier reviewed (changes requested):
Yes, annoying; maybe it can be fixed this way for 32bit only.

aaryapatil commented:
The change still doesn’t work for 32 bit. Would it be reasonable to do a skip?

dhomeier commented on astropy/timeseries/tests/test_downsample.py:
```diff
@@ -33,38 +39,123 @@ def test_timeseries_invalid():
+    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5]], names=['a'])
+    ts_units = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5] * u.count], names=['a'])
 
     down_1 = aggregate_downsample(ts, time_bin_size=1*u.second)
```
```suggestion
    if np.finfo(float).dtype == np.float32:
        time_bin_incr = (1 - 1e-6) * u.s
    else:
        time_bin_incr = 1 * u.s

    down_1 = aggregate_downsample(ts, time_bin_size=1*time_bin_incr)
```
Sorry if I haven’t been clear; my computer has all but blown up and it’s difficult to see on a phone what exactly your latest changes were or edit in the GH interface. But what I meant was, don’t change `INPUT_TIME` at all but instead set time_bin_size this way, and then with multiples 2, 3 etc.
If that doesn’t work, yes, skipping is ok with me.

aaryapatil replied:
Aah, that makes sense. I thought you were suggesting changes the `INPUT TIMES`.

aaryapatil replied:
That still doesn't work 😞 . I think I will skip the test.

aaryapatil commented on astropy/timeseries/tests/test_downsample.py:
```diff
@@ -33,38 +36,125 @@ def test_timeseries_invalid():
+    assert_equal(down_nbins.time_bin_start.isot, ['2016-03-22T12:30:31.000', '2016-03-22T12:30:33.000'])
 
 
+@pytest.mark.skipif(np.finfo(float).dtype == np.float32,
```
This is surprisingly not getting skipped in the 32 bit run.

dhomeier replied:
> This is surprisingly not getting skipped in the 32 bit run. Maybe that has been the issue all this while.

Strange. Yes, would be interesting if the bin sizes would work with `sys.maxsize` as well, but I am leaving that up to you to decide; I’ll approve it either way.

aaryapatil replied:
I'm fine either way. Might be easier to merge this and figure out how to change tests later for 32 bit.

dhomeier replied:
> I'm fine either way. Might be easier to merge this and figure out how to change tests later for 32 bit.

If
```python
    if sys.maxsize <= 2**32:
        time_bin_incr = (1 - 1e-6) * u.s
```
worked right away, it probably would be worth having a functional test, but if not, I'd just revert to the current HEAD and get that merged.

aaryapatil replied:
Great I'll add that in!

aaryapatil replied:
It works better than before: The failing test shows a difference in only the first element. But it still fails.

dhomeier commented:
Looks better, but apparently the first bin ended up empty now. Maybe the single `time_bin_start` needs to be decremented a ns as well.

dhomeier commented:
I think @astrofrog would just have to confirm that the overlapping bins are addressed in a satisfactory manner.

aaryapatil commented:
I could work that into the code. Maybe when `time_bin_start` is scalar, we should just reduce it by a `ns` to allow some leeway for the start time.

dhomeier commented:
> Maybe when `time_bin_start` is scalar, we should just reduce it by a `ns` to allow some leeway for the start time.

Exactly. Except here it even seemed to fail with the default start of `ts.time[0]`, so it probably will have to be set explicitly (sorry, again don’t have a computer to test myself).

aaryapatil commented:
No worries! I will test it. Are you suggesting I manually set the start time for these tests for the 32 bit case?
I will try passing `time_bin_start ` explicitly in the tests as follows:
```
if sys.maxsize <= 2**32:
    time_bin_start = ts.time[0] + 1*u.ns
else:
    time_bin_start = None
```

dhomeier commented:
Thanks for your patience!
I think this would have been ready to merge last week and add the 32-bit cleanup as a bug fix, but getting it in fully working is much preferable.
I think it’s also good to keep a track of these sometimes qualitatively different behaviour in 32-bit Time objects.
Only waiting for word from @taldcroft now.
(And I guess since this is milestoned to 5.0 now it’s past the time for @eteq to set up a backport commit).

dhomeier commented:
Pinging @plim and @eteq about a merge

aaryapatil commented:
Any updates @bsipocz @astrofrog?
P.S. I will fix the merge conflict.

dhomeier commented:
> Any updates @bsipocz @astrofrog?
> P.S. I will fix the merge conflict.

4.3 is basically out; pinging @eteq again if this means leaving the milestone at 5.0 or setting to 4.3.1 or something else (especially in light of the discussion on new features in LTS in https://github.com/astropy/astropy/pull/11860#issuecomment-865278611 ff.).

@astrofrog I had some further thoughts about the overlapping and in particular the "marginally overlapping" bins (i.e. data points exactly on the `time_bin_start` times). At the time it seemed natural that the desired default behaviour would be to assign a point with
`ts.time[i] == time_bin_start[j]`
to bin `j`; but actually if we assume `ts.time[i]` to represent the mean time of an observation (the perennial issue of sample centre vs. sample borders...), it would be just as natural to assign this with weight 0.5 each to ` time_bin_start[j-1]` **and** ` time_bin_start[j]`. So I think there is justification for a new option to control how these points are handled, but that could actually be extended to cases where `ts.time[i]` is not exactly on the boundary (and would be distributed with different weights). So, plenty of tasks for a follow-up on this PR.

bsipocz commented:
> 4.3 is basically out; pinging @eteq again if this means leaving the milestone at 5.0 or setting to 4.3.1 or something else (especially in light of the discussion on new features in LTS in #11860 (comment) ff.).

The milestone here should be 5.0. Absolutely no new features into bugfix (e.g. into 4.3.1). And the comment for LTS was about the preference of no new, significantly big APIs, e.g a new module, a major new feature with complex API, etc. So this is all in the clear.

dhomeier commented:
> The milestone here should be 5.0. Absolutely no new features into bugfix (e.g. into 4.3.1). And the comment for LTS was about

OK. This was in the feature freeze phase of 4.3 when the last call for reviews came out, and it was not quite clear to me what the status of 4.3.1 would be.

@aaryapatil I think you just need to pull in the doc changes from main; will see if the doctests are passing then.

bsipocz commented:
> OK. This was in the feature freeze phase of 4.3 when the last call for reviews came out, and it was not quite clear to me what the status of 4.3.1 would be.

It's the merge that matters. If it misses FF then it's next release, especially when it's not just FF that passed but when we're in RC testing.  A bugfix release by definition should never have any new features.

dhomeier commented:
@astrofrog @bsipocz any news from the maintainers' side on this before we need to start worrying about the 5.0 feature freeze?

aaryapatil commented:
@astrofrog @bsipocz Is there an update regarding this PR? Hopefully this goes into 5.0!

keflavich reviewed (commented):
I gave the changes a quick skim, but I don't feel that I can give a high-quality review right now; there's a lot changed and the underlying code is beyond my experience.  The interface looks clean enough to me, and I think the overall logic looks OK, but I haven't carefully read the implementation details.  Consider this a 'tentative approval' - if you're confident and just needed a maintainer's approve to go green, I can provide that.

dhomeier commented:
Thanks a lot for your review @keflavich! Yes, primarily we are looking for a second opinion to allow this to be merged before potentially missing another feature freeze. Basically responding to @pllim's concern https://github.com/astropy/astropy/pull/11266/#issuecomment-836718911 about merging without maintainer approval.
The specific issue left open is @astrofrog's request https://github.com/astropy/astropy/pull/11266/#pullrequestreview-611744023 for a sensible handling of overlapping bins.
My take is that the final word might not have been spoken on the treatment of points exactly on the (`time_bin_start`) edge of a bin as discussed in https://github.com/astropy/astropy/pull/11266/#issuecomment-865942363, but that should not hold up this PR any more.

aaryapatil commented:
Thanks @keflavich! I am pretty confident that this does the intended job of adding a `custom binning` feature. As @dhomeier mentioned, there is definitely room for improvement. However, the handling of overlapping bins, specifically the assignment of points exactly at the start of a bin edge, warrants some discussion and can be treated as a special case to be dealt with later. I am happy to do that in a separate PR. 

All in all, I don't think improvements should stop this PR from merging since it introduces the basis functionality of `custom bins`.

dhomeier commented:
Thanks for merging @hamogu! I definitely would not have wanted this to slip through the next feature freeze.
